core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java

```diff
@@ -151,13 +151,17 @@ public BaseUpdatePartitionSpec addField(String name, Term term) {
     transformToAddedField.put(validationKey, newField);
 
     PartitionField existingField = nameToField.get(newField.name());
-    if (existingField != null) {
+
+    if (existingField != null && !deletes.contains(existingField.fieldId())) {
       if (isVoidTransform(existingField)) {
         // rename the old deleted field that is being replaced by the new field
         renameField(existingField.name(), existingField.name() + "_" + existingField.fieldId());
       } else {
         throw new IllegalArgumentException(String.format("Cannot add duplicate partition field name: %s", name));
       }
+    } else if (existingField != null && deletes.contains(existingField.fieldId())) {
+      renameFieldInternal(existingField.name(), existingField.name() + "_" + existingField.fieldId(),
+              true);
```

Comment:
It would be better to wrap all the arguments on a single line rather than have just the last one on its own.

```java
      renameFieldInternal(
          existingField.name(), existingField.name() + "_" + existingField.fieldId(), true);
```